src/hevm/src/EVM/SMT.hs

```diff
@@ -824,24 +824,20 @@ expandExp base expnt
     pure $ "(* " <> b `sp` n <> ")"
 
 -- | Concatenates a list of bytes into a larger bitvector
-concatBytes :: [Expr Byte] -> State BuilderState Text
-concatBytes [b] = exprToSMT b
-concatBytes (hd : tl) = do
-  eHd <- exprToSMT hd
-  eTl <- concatBytes tl
-  pure $ "(concat " <> eHd `sp` eTl <> ")"
-concatBytes [] = error "cannot concat an empty list of bytes" -- TODO: use nonempty here?
+concatBytes :: NonEmpty (Expr Byte) -> State BuilderState Text
+concatBytes bytes = foldM wrap "" $ NE.reverse bytes
+  where
+    wrap inner byte = do
+      byteSMT <- exprToSMT byte
+      pure $ "(concat " <> byteSMT `sp` inner <> ")"
 
 -- | Concatenates a list of bytes into a larger bitvector
-writeBytes :: Expr EWord -> [Expr Byte] -> Expr Buf -> State BuilderState Text
-writeBytes _ [] buf = exprToSMT buf
-writeBytes idx [b] buf = do
-  eBuf <- exprToSMT buf
-  eIdx <- exprToSMT idx
-  eByte <- exprToSMT b
-  pure $ "(store " <> eBuf `sp` eIdx `sp` eByte <> ")"
-writeBytes idx (hd : tl) buf = do
-  eHd <- exprToSMT hd
-  eIdx <- exprToSMT idx
-  eTl <- writeBytes (add idx (Lit 1)) tl buf
-  pure $ "(store " <> eTl `sp` eIdx `sp` eHd <> ")"
+writeBytes :: [Expr Byte] -> State BuilderState Text
```

Comment:
Not sure if I'm reading this write, but it seems less general than the old `writeBytes`, which could write a region of bytes at an arbitrary index onto an arbitrary buffer? This version seems to just write them onto an `EmptyBuf`?

Reply:
Right, I didn't know how general this should be. I can add an argument back to `writeBytes` and pass an `EmptyBuf` at the call site.

Reply:
Changed